### ISSUE TITLE
Traceback during update_stores in default terminology project

### ISSUE THREAD

**Issue by dwaynebailey** [translate/pootle#3563]
In a default install following: ``migrate`` + ``initdb``.  During a run of ``update_stores`` across the two default projects I get the following traceback.  The only other change was moving ``PODIRECTORY`` to ``po2``

```pytb
 pootle.core.mixins.treeitem.update_cache(<Store: /ln/terminology/GM_ln_20090916.po>, set(['get_last_updated', 'get_fuzzy_wordcount', 'get_checks', 'get_total_wordcount', 'get_translated_wordcount', 'get_mtime', 'get_suggestion_count', 'get_last_action'])) from default
4665172d-b561-4203-bc7a-30a946852674
Failed 3 minutes ago

Traceback (most recent call last):
  File "/Users/dwayne/dev/envs/pootle/lib/python2.7/site-packages/rq/worker.py", line 479, in perform_job
    rv = job.perform()
  File "/Users/dwayne/dev/envs/pootle/lib/python2.7/site-packages/rq/job.py", line 466, in perform
    self._result = self.func(*self.args, **self.kwargs)
  File "/Users/dwayne/dev/pootle-dwayne/pootle/core/mixins/treeitem.py", line 506, in update_cache
    instance._update_cache(keys)
  File "/Users/dwayne/dev/pootle-dwayne/pootle/core/mixins/treeitem.py", line 473, in _update_cache
    p._update_cache(keys)
  File "/Users/dwayne/dev/pootle-dwayne/pootle/core/mixins/treeitem.py", line 473, in _update_cache
    p._update_cache(keys)
  File "/Users/dwayne/dev/pootle-dwayne/pootle/core/mixins/treeitem.py", line 471, in _update_cache
    self.update_cached(key)
  File "/Users/dwayne/dev/pootle-dwayne/pootle/core/mixins/treeitem.py", line 63, in _statslog
    result = function(instance, *args, **kwargs)
  File "/Users/dwayne/dev/pootle-dwayne/pootle/core/mixins/treeitem.py", line 289, in update_cached
    self.set_cached_value(name, self._calc(name, from_update=True))
  File "/Users/dwayne/dev/pootle-dwayne/pootle/core/mixins/treeitem.py", line 224, in _calc
    return self._calc_last_updated(from_update)
  File "/Users/dwayne/dev/pootle-dwayne/pootle/core/mixins/treeitem.py", line 199, in _calc_last_updated
    key=lambda x: x['creation_time'] if 'creation_time' in x else 0
  File "/Users/dwayne/dev/pootle-dwayne/pootle/core/mixins/treeitem.py", line 199, in <lambda>
    key=lambda x: x['creation_time'] if 'creation_time' in x else 0
TypeError: argument of type 'NoneType' is not iterable
```

**Comment by julen** [comment]
We've seen exceptions like these in our server. Any clues @ta2-1?

**Comment by dwaynebailey** [comment]
@julen Wow, I'm getting a tonne of them while rerunning the tests on  #3545 - they didn't exist when I first did that investigation,

**Comment by ta2-1** [comment]
@dwayne, and you moved `po` folder to `po2` in filesystem, did you? 
what did you do with cached data in Redis?

**Comment by dwaynebailey** [comment]
@ta2-1 it was a new install, I just reconfigure PODIRECTORY to be different from the default of ``po``. I'm using memcached still for these caches.  I don't think the memcached would impact it and I wouldn't think stale cache would either

**Comment by ta2-1** [comment]
@dwaynebailey
Hmmm, probably I misunderstood what a new install mean.
There is `po` directory in our repo (https://github.com/translate/pootle/tree/master/pootle/po).
I do following to get a new install:
- change database name to an empty database (or delete existing file if sqlite is used);
- run `./manage.py syncdb`, `./manage.py migrate`, `./manage.py initdb`
- two projects are created: `tutorial` and `terminology`
- if PODIRECTORY exists all data from `.po` files is loaded to DB 
Thus as I understand: if you change PODIRECTORY` you should set it to existing folder with `.po` files.

Another thing:
We got memcached evictions when we used memcached on production i.e. cached values were removed because of memcached size limit. 
If any of cached values are removed you get errors in RQ jobs.

Re this issue:
Do you get this traceback in RQ job?
Do you have DEBUG=True in your settings?

**Comment by dwaynebailey** [comment]
@ta2-1 Sorry the part about ``PODIRECTORY`` is confusing things.  @julen simply wanted me to confirm that if I changed ``PODIRECTORY`` the migrations would work from a new install.  So I simply created po2 and copied into it ``tutorial`` and ``terminology`` projects and changed ``PODIRECTORY`` to suite that.

Yes, the install is as you outline except my minor ``PODIRECTORY`` change and that there is no ``syncdb`` anymore.

Ahhha the tracebacks are within RQ jobs.  Sorry I didn't add that to the initial report, thought I had.  I can't recall if I had DEBUG=True, looks like I don't.

**Comment by julen** [comment]
> * run ./manage.py syncdb, ./manage.py migrate, ./manage.py initdb

@ta2-1 note that `syncdb` is not necessary anymore, `migrate` covers everything you need in terms of DB syncing, including migrations of course. I have updated the docs accordingly.

**Comment by dwaynebailey** [comment]
@ta2-1 got this again and thought it a good time to get the minimum replications.

1. I have a checkout of #3584 though I assume ``master`` would also do the same
2. I'm testing on sqlite I don't think this would be an issue, but it could be.  The default pootle setup for sqlite should be enough.
3. Ran ``./manage.py migrate`` and ``./manage.py initdb``
4. Redis and 1 rqworker are in place
5. Ran ``./manage.py runserver --insecure`` with ``DEBUG=False``
6. Run ``./manage.py update_stores`` they start running and we start seeing the [above listed traceback](https://github.com/translate/pootle/issues/3563#issue-55904920) and the jobs go into the failed queue.

**Comment by ta2-1** [comment]
@dwaynebailey I wrote in #pootle-dev that we have incorrect default cache settings here:
https://github.com/translate/pootle/blob/master/pootle/settings/20-backends.conf#L29-L39
we require infinite TIMEOUT for stats cache.

**Comment by dwaynebailey** [comment]
@ta2-1 I saw your comments on the channel and made those changes.  The rest of my notes are for someone hitting this in future.

I made the cache setup changes to ensure that there is a ``stats`` cache with no ``TIMEOUT`` and a ``default`` with more normal timeout values.  That initially didn't seem to fix it. Then I stepped carefully through the setup.  I killed Redis and made sure the backup DB, ``dump.rdb`` was also deleted.  I then carefully did a new setup and made sure Redis and rqworker were running before any Pootle command were issued.  With that careful setup it worked perfectly.

So it seems the risks could be:

- Redis not running
- The remains of a corrupted Redis DB
- Incorrect cache ``TIMEOUT``

I'm not sure we need to solve the specific traceback in this issue. So I'm marking this as invalid.  Though I am concerned that if the setup isn't 100% right then we run the risk of getting broken stats or worse that we can't get back out without some risky 'fix' to a running system.

I think we do need to do the following:

- Make sure documentation for the Redis cache is correct (#3351)
- Ensure that we by default have a stats cache with correct timeout (#3585)
